### DISCUSSION TITLE
No way to access DDR5 SDRAM information.

### DISCUSSION THREAD

**Discussion by roke-julian-lockhart** [hardinfo2/hardinfo2#63] [Ideas]
I'm using DDR5 SDRAM:

1.	```sh
	#!/usr/bin/env sh
	sudo dmidecode -t 'memory' | grep -e 'DDR'
	```

	[^1]

[^1]: [`superuser.com/review/suggested-edits/1260313`](https://superuser.com/review/suggested-edits/1260313#:~:text=sudo%20dmidecode%20%2Dt%20'memory'%20%7C%20grep%20%2De%20'DDR')
	
1.	> ```YAML
	> Type: DDR5
	> ```

However, `hardinfo2` doesn't provide instructions on how to access information for it:

![Screenshot_20240812_203420](https://github.com/user-attachments/assets/93f070a2-9b17-44d0-8c5c-46bcd12cc76d)

**Comment by hwspeedy** [comment]
Please put DDR5 in a todo list in discussions - have you tried ee1004 and checked up on spd code/spec for DDR5 vs DDR4?
Maybe also check if there are updates to the spd decode package...

Please use discussion for ideas to gather people to create some code for a PR
Please keep issues clean with only real issues to avoid a huge backlog that no one wants to touch, thanx.

**Reply by roke-julian-lockhart** [reply]
> have you tried ee1004 and checked up on spd code/spec for DDR5 vs DDR4?

@hwspeedy, I've not tried that. I didn't intend to add a kernel module that I didn't appear to need. Might it support SDDR5?

> Maybe also check if there are updates to the spd decode package...

What's the exact package name?

**Reply by hwspeedy** [reply]
In the source code are 2 files with spd (Serial Presence Device) - located on the DIMMs.

/*
 * spd-decode.c, spd-vendors.c
 * Copyright (c) 2010 L. A. F. Pereira
 * modified by Ondrej Čerman (2019)
 * modified by Burt P. (2019)
 *
 * Based on **decode-dimms.pl**
 * Copyright 1998, 1999 Philip Edelbrock <phil@netroedge.com>
 * modified by Christian Zuckschwerdt <zany@triq.net>
 * modified by Burkart Lingner <burkart@bollchen.de>
 *

**Reply by roke-julian-lockhart** [reply]
> https://github.com/hardinfo2/hardinfo2/discussions/63#discussioncomment-10318574

@hwspeedy, per:

1.	https://github.com/hardinfo2/hardinfo2/blob/65b5768528d9a97dbca46f843a628f29331509b3/modules/devices/spd-decode.c#L48-L52 and
1.	https://github.com/hardinfo2/hardinfo2/blob/65b5768528d9a97dbca46f843a628f29331509b3/modules/devices/spd-decode.c#L56-L59

...it doesn't appear to support DDR5.

**Reply by hwspeedy** [reply]
@ocerman - is active sometimes - maybe he has input to DDR5 spd decoding.
Availability of Linux DDR5 systems are still low, so not that many that can hack and test yet.

**Reply by hwspeedy** [reply]
decode-dimms is part of i2c-tools package, that seams to be part of kernel tools. Last updated 2021-07.
https://www.kernel.org/pub/software/utils/i2c-tools/

which does not have DDR5 support yet:
https://access.redhat.com/solutions/7053692

The SPD is using I3C with a hub, so some changes and hacking needed.

**From:**
https://www.kingston.com/en/blog/pc-performance/ddr5-overview#:~:text=and%20reduces%20noise.-,SPD%20Hub,the%20internal%20bus%20from%20external.

**SPD hub** 
DDR5 utilises a new device that integrates the Serial Presence Detect (SPD) EEPROM with additional hub features, manages access to the external controller and decouples the memory load on the internal bus from external.

**SPD INFO**
https://www.unisemicon.com/uploadfile/2023/0228/20230228104504760.pdf

**Reply by t-8ch** [reply]
> have you tried ee1004

For DDR5 the spd5118 driver is necessary.
It will be part of Linux 6.11.

**Reply by hwspeedy** [reply]
Thanx, So we need to use spd5118 and add logic to dmi_memory.c

https://github.com/hardinfo2/hardinfo2/commit/34f932e74a7abc8271e503530148e660402b9d2a

@ocerman @bp0 - Any of you have the knowledge, time and want to help getting dmi memory running with DDR5? @RokeJulianLockhart has a Linux DDR5 system and can maybe also help?

**Reply by t-8ch** [reply]
I have posted a DDR5 eeprom dump here:
https://lore.kernel.org/lkml/cf9d752e-0137-4a6d-85d3-fbe69293a43e@t-8ch.de/

**Reply by roke-julian-lockhart** [reply]
> https://github.com/hardinfo2/hardinfo2/discussions/63#discussioncomment-10318650

@hwspeedy, as [this Discourse thread](https://community.frame.work/t/responded-amd-ryzen-7040-decode-dimms-no-ddr5-modules-found/39704/11?u=rokejulianlockhart) explains, I've filed a [a ticket at the kernel's Bugzilla instance](https://bugzilla.kernel.org/show_bug.cgi?id=219155#c0:~:text=Per%20https://github.com/hardinfo2/hardinfo2/discussions/63%23discussioncomment%2D10322689%2C%20there%20has%20been,implement%20it%20when%20it's%20merged) so that support can be tracked. You may want to subscribe.

**Reply by ValeZAA** [reply]
I applied https://lore.kernel.org/all/20241114-decode-ddr5-v1-5-0ed2db8ef30f@outlook.com.au/T/ and compiled the new i2c-tools's decode-dimms, but it still does not work, /sys/bus/i2c/drivers/spd5118/0-0053/eeprom is empty. Does that mean i2c is not supported, i3c only?

**Reply by hwspeedy** [reply]
@ValeZAA 

The DDR5 modules I tested was also using i2c and was in /sys/bus/i2c/drivers/spd5118/... like yours. 
The spd5118 module autoloaded and 1KB data was available.
Tested on Fedora41 nightly builds with hardinfo2.

You are probably missing some kernel patches - try a new fedora rawhide and see it works with hardinfo2.

PS: Don't see any DDR5 updates for i2c-tools - https://git.kernel.org/pub/scm/utils/i2c-tools/i2c-tools.git/log/

**Reply by ValeZAA** [reply]
I have 6.14.3 kernel. What kernel patches?

I said I compiled my own i2c-tools with those patches i linked to.

**Reply by hwspeedy** [reply]
If kernel module spd5118 does not work for you - it is the Linux kernel maintaining this driver - right?

You should go to the kernel mailing list with questions about non merged/released i2c-tools - decode-dimms. - right?

As suggested try to use hardinfo2 on Fedora Rawhide to see if you have success like this to avoid wasting the kernel developers time:
![image](https://github.com/user-attachments/assets/3b9673e0-6c0e-4783-b69a-3e324ce3a597)

Have a nice day.

**Reply by hwspeedy** [reply]
@ValeZAA
Why are you answering a different old post now? https://github.com/hardinfo2/hardinfo2/discussions/63#discussioncomment-13017899
Are you spamming on purpose?

Above is the screenshot - reading SPD via i2c on 51+53 as populated in system.

**Reply by ValeZAA** [reply]
In my case it does not show i2c entry. It says unknown like on screenshot below. I make install'ed new decode dimms https://github.com/hardinfo2/hardinfo2/discussions/63#discussioncomment-10776025

Latest hardinfo2. Basically the same issue as this. Sigh:

https://lkml.indiana.edu/hypermail/linux/kernel/2406.2/03497.html

Thankfully i3c patches are there. https://lore.kernel.org/all/?q=Guenter+Roeck+eeprom+

**Comment by hwspeedy** [comment]
Fedora RawHide now is on 6.11 and has the spd5118 module.

Anyone with a DDR5 machine wants to "hack" and get SPD working for DDR5 modules?

**Comment by hwspeedy** [comment]
@RokeJulianLockhart @topazus

Hardinfo2 - SPD DDR5 on kernel 6.11

Added initial DDR5 SPD support:
https://github.com/hardinfo2/hardinfo2/commit/c07dd3840d71fde80a8fe8a49322191e1ff88d4a

Please test git->master (unstable) on your DDR5 machines with kernel 6.11 and spd5118 kernel module, thanx.

**Reply by roke-julian-lockhart** [reply]
> https://github.com/hardinfo2/hardinfo2/discussions/63#discussioncomment-10773475

@hwspeedy, I can't test `master`, but I can wait for it to be compiled for Fedora 40 (which if you have submitted it to Koji, shouldn't be long).

<details>

1.	> ![Screenshot_20240927_171526](https://github.com/user-attachments/assets/347a4148-4881-41a5-9c4a-c2741058d99a)

1.	> ```log
	> PS /home/RokeJulianLockhart> sudo modprobe spd5118        
	> [sudo] password for RokeJulianLockhart: 
	> modprobe: FATAL: Module spd5118 not found in directory /lib/modules/6.10.11-200.fc40.x86_64
	> ```

</details>

Anyway, thank you. I'm surprised you managed it without support in `decode-dimms`.

**Reply by hwspeedy** [reply]
Fedora 41 will be kernel 6.11

Fedora 40 is only kernel 6.10 which do not have the spd5118.

So you have to wait for some time...

**Reply by roke-julian-lockhart** [reply]
> https://github.com/hardinfo2/hardinfo2/discussions/63#discussioncomment-10774373

@hwspeedy, so it's also a change on the kernel's end? If so, would it be worth closing https://bugzilla.kernel.org/show_bug.cgi?id=219155#c0?

**Reply by hwspeedy** [reply]
The perl script should be updated as well and RedHat is working on it. https://access.redhat.com/solutions/7053692

**Comment by topazus** [comment]
Tested with the latest commit
![image](https://github.com/user-attachments/assets/d45d1e90-0c8e-4d5b-b292-4cb3e299a136)

**Reply by ValeZAA** [reply]
On your screenshot it does not read SPD either 50 and 52 either? That data you showed is not SPD, it is from Intel DMI, not from I3C/I2C.

**Comment by hwspeedy** [comment]
It is probably the location i2c vs i3c or eeprom naming with the spd5118 driver.

Please log these outputs:
ls -l /sys/bus/i2c/devices/15-0050/
ls -l /sys/bus/i3c/devices/15-0050/

Thanx

**Reply by topazus** [reply]
> ls -l /sys/bus/i2c/devices/15-0050/
> ls -l /sys/bus/i3c/devices/15-0050/

```bash
❯ ls -l /sys/bus/i2c/devices/
total 0
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 13-0050 -> ../../../devices/pci0000:00/0000:00:1f.4/i2c-13/13-0050
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 13-0052 -> ../../../devices/pci0000:00/0000:00:1f.4/i2c-13/13-0052
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 i2c-0 -> ../../../devices/pci0000:00/0000:00:02.0/i2c-0
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 i2c-1 -> ../../../devices/pci0000:00/0000:00:02.0/i2c-1
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 i2c-10 -> ../../../devices/i2c-10
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 i2c-11 -> ../../../devices/i2c-11
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 i2c-12 -> ../../../devices/i2c-12
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 i2c-13 -> ../../../devices/pci0000:00/0000:00:1f.4/i2c-13
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 i2c-2 -> ../../../devices/pci0000:00/0000:00:02.0/i2c-2
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 i2c-3 -> ../../../devices/pci0000:00/0000:00:02.0/i2c-3
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 i2c-4 -> ../../../devices/pci0000:00/0000:00:02.0/i2c-4
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 i2c-5 -> ../../../devices/pci0000:00/0000:00:02.0/i2c-5
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 i2c-6 -> ../../../devices/pci0000:00/0000:00:02.0/i2c-6
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 i2c-7 -> ../../../devices/pci0000:00/0000:00:02.0/i2c-7
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 i2c-8 -> ../../../devices/pci0000:00/0000:00:02.0/i2c-8
lrwxrwxrwx. 1 root root 0 Sep 29 10:22 i2c-9 -> ../../../devices/pci0000:00/0000:00:02.0/drm/card1/card1-eDP-1/i2c-9

❯ ls -l /sys/bus/i2c/devices/15-0050/
ls -l /sys/bus/i3c/devices/15-0050/
ls: cannot access '/sys/bus/i2c/devices/15-0050/': No such file or directory
ls: cannot access '/sys/bus/i3c/devices/15-0050/': No such file or directory

```

**Reply by ValeZAA** [reply]
what does `ls -la /sys/bus/i2c/drivers/spd5118/*-005*/eeprom` say? then cat /sys/bus/i2c/drivers/spd5118/x-005x/eeprom (put what ls found)

**Comment by hwspeedy** [comment]
@topazus 
along with above outputs - I guess it's a wrong bus - should be i3c.

Please also try latest where the driver is on i3c bus.
https://github.com/hardinfo2/hardinfo2/commit/0b666c056d862bf7473f68cb346aa9593dc0a0b0

Thanx

**Reply by t-8ch** [reply]
Spd5118 does not support i3c

**Comment by hwspeedy** [comment]
@t-8ch 
What is the path for eeprom?

**Reply by t-8ch** [reply]
```
/sys/devices/pci0000:00/0000:00:14.0/i2c-20/20-0050/20-0050/nvmem
/sys/devices/pci0000:00/0000:00:14.0/i2c-20/20-0051/20-0051/nvmem
```

**Reply by hwspeedy** [reply]
Thanx,
so first DIMM SPD eeprom can be found at /sys/bus/i2c/drivers/spd5118/20-0050/nvmem for your system - right?
/sys/bus/i2c/drivers/spd5118/15-0050/nvmem for topazus system.

**Reply by t-8ch** [reply]
You should use /sys/bus/nvmem/devices and check for the driver in use instead of fiddling with the full paths.

**Reply by hwspeedy** [reply]
Why are we moving the SPD eeprom to nvmem and adding an extra layer?

**a new DDR5 SPD test for @topazus**
 - This is using the /sys/bus/i2c 
 - I know from above picture from topazus that spd5118 can be found at /sys/bus/i2c/drivers/spd5118
 - So change is to use nvmem instead of eeprom for reading the spd eeprom.

https://github.com/hardinfo2/hardinfo2/commit/2b664d61e2a5f27e5f84ed99878b77fb85a54517

**Reply by t-8ch** [reply]
Because the eeprom is in fact an nvmem device.
The eeprom interface is only the backwards compat shim.

Looking at the details of the device path is unreliable.
They can change for all kinds of reasons.
Also you can't correlate i2c devices to DMI memory slots unfortunately.

**Reply by hwspeedy** [reply]
All eeproms are non volatile memory. 
As I read the kernel doc for nvmem it adds a register map - like when you have a NVMEM device and want to only give access to the MAC address etc.

> Looking at the details of the device path is unreliable.
> They can change for all kinds of reasons.

Fun stuff - But it's probably not overengineered - DCDC is now on module, sleep, Fault detection, RGB, etc. - it has to maybe do a lot more than just comply to jedec JESD400-5C..

It seams the i2c addresses for DDR5 spd5118 devices is following the known scheme for DIMM slots 0x50->
- we use a best match algorithm between DMI and SPD (low stuff, but seams to work well)

We will have to see the results with DDR5, but first we need to could read the eeproms (Perhaps the latest patch could do that) - The basic parsing of DDR5 spd eeprom works with your DDR5 dump.

**Reply by t-8ch** [reply]
> All eeproms are non volatile memory. As I read the kernel doc for nvmem it adds a register map - like when you have a NVMEM device and want to only give access to the MAC address etc.

This is one usecase. But here the nvmem subsystem is only used in its basic form. For discovery and API uniformity.

> It seams the i2c addresses for DDR5 spd5118 devices is following the known scheme for DIMM slots 0x50->

Up to 0x57, 8 DIMMs in total.
But there can be more than 8 DIMMs per system.
On Intel chipset you will get multiple busses.
On Piix4 it will just be truncated.

>     * we use a best match algorithm between DMI and SPD (low stuff, but seams to work well)

FYI the kernel explicitly does not evem try to do such a mapping.

> We will have to see the results with DDR5, but first we need read the eeproms - The basic parsing works with your DDR5 dump.

I don't get this sentence.

**Reply by hwspeedy** [reply]
> Up to 0x57, 8 DIMMs in total.
> But there can be more than 8 DIMMs per system.
> 

Server SPD have sofar not been accessible via I2C - they have a BMC that handles that.
DMI works for most's servers - which is fine - nothing to tune, check or test on a server.

> I don't get this sentence.

Sorry it's friday :)
We will have to see how good the DDR5 SPD<->DMI matching is, when topazus runs the latest commits and it hopefully works - little difficult to bring up without a Linux 6.11 + DDR5 System.
And no the matching should not be in the kernel - it's guessing.

Thanx for your help,

**Reply by t-8ch** [reply]
> Server SPD have sofar not been accessible via I2C - they have a BMC that handles that.
> DMI works for most's servers - which is fine - nothing to tune, check or test on a server

I tested this on a 32 slot server with 16 slots populated.
4 DIMMs were accessible through i2c.
Only 4 because of the Piix4 issue mentioned above.

This was DDR4, but I don't see why DR5 should be different.

**Comment by topazus** [comment]
build with https://github.com/hardinfo2/hardinfo2/commit/78cc875198c0006993e47976e652538b4a5e771b
![image](https://github.com/user-attachments/assets/9648572a-b221-467f-9575-59ac85570e71)

**Comment by hwspeedy** [comment]
Thanx topazus, It's still not okay.

Can you send me your first SPD eeprom?
 - according to above info, it should be at /sys/bus/i2c/drivers/spd5118/13-0050/nvmem
 - cp  /sys/bus/i2c/drivers/spd5118/13-0050/nvmem eeprom-DDR5-1.bin

**Comment by hwspeedy** [comment]
spd5118 is still supplying an eeprom on i2c interface - just like all other SPD eeproms. Fixed https://github.com/hardinfo2/hardinfo2/commit/4b27e9537cc6be5a92de07236ae9b83f3c254e7d

So master now works with DDR5 spd decoding, matching and showing basic DDR5 SPD information:
![hardinfo2fc41](https://github.com/user-attachments/assets/9e32dc1c-80b7-4b0b-a54c-d395d2e5a542)

note: FC41 nightly build from 2024-09-30 auto loads spd5118.
